certbot-ci/certbot_integration_tests/utils/dns_server.py

```diff
@@ -0,0 +1,136 @@
+    use for each pytest node.
+
+    At this time, DNSServer should only be used with a single node, but may be expanded in
+    future to support parallelization.
+    """
+
+    BIND_DOCKER_IMAGE = 'internetsystemsconsortium/bind9:9.16'
+    BIND_BIND_ADDRESS = ('127.0.0.1', 45953)
+
+    # A DNS message which is a query for '. IN A' transaction ID 0xe785. This is used by
+    # _wait_until_ready to check that BIND is responding without depending on dnspython.
+    BIND_TEST_QUERY = bytearray.fromhex('0028e7850120000100000000000100000100010000'
+                                        '29100000000000000c000a00083ad084e525a28702')
+
+
+    def __init__(self, nodes, show_output=False):
+        """
+        Create an DNSServer instance.
+        :param list nodes: list of node names that will be setup by pytest xdist
+        :param bool show_output: if True, print the output of the DNS server
+        """
+
+        self.bind_root = tempfile.mkdtemp()
+
+        # Same port to every node. We are only pretending to support parallelization.
+        self.dns_xdist = {
+            node: DNSServer.BIND_BIND_ADDRESS[1] for node in nodes
+        }
+        self.dns_xdist['address'] = DNSServer.BIND_BIND_ADDRESS[0]
+
+        # Unfortunately the BIND9 image forces everything to stderr with -g and we can't
+        # modify the verbosity.
+        self._output = sys.stderr if show_output else open(os.devnull, 'w')
+
+    def start(self):
+        """Start the DNS server"""
+        try:
+            self._configure_bind()
+            self._start_bind()
+        except:
+            self.stop()
+            raise
+
+    def stop(self):
+        """Stop the DNS server, and clean its resources"""
+        try:
+            self.process.terminate()
+            self.process.wait()
+        except:
+            pass
+        finally:
+            shutil.rmtree(self.bind_root)
+            if self._output != sys.stderr:
+                self._output.close()
+
+    def _configure_bind(self):
+        """Configure the BIND9 server based on the prebaked configuration"""
+        bind_conf_src = resource_filename('certbot_integration_tests', 'assets/bind-config')
+        for dir in ['conf', 'zones']:
+          shutil.copytree(os.path.join(bind_conf_src, dir), os.path.join(self.bind_root, dir))
+
+    def _start_bind(self):
+        """Launch the BIND9 server as a Docker container"""
+        addr_str = '{}:{}'.format(DNSServer.BIND_BIND_ADDRESS[0], DNSServer.BIND_BIND_ADDRESS[1])
+        self.process = subprocess.Popen([
+          'docker', 'run', '--rm',
+          '-p', '{}:53/udp'.format(addr_str),
+          '-p', '{}:53/tcp'.format(addr_str),
+          '-v', '{}/conf:/etc/bind'.format(self.bind_root),
+          '-v', '{}/zones:/var/lib/bind'.format(self.bind_root),
+          DNSServer.BIND_DOCKER_IMAGE
+        ], stdout=self._output, stderr=self._output)
+
+        assert self.process.poll() is None
```

Comment:
The `assert` keywork can be disabled if the Python interpreter is not run in debug mode (`python -O`). I think an explicit raise will be more universal.